### DISCUSSION TITLE
Add Patent Entry

### DISCUSSION THREAD

**Discussion by RashKash103** [rendercv/rendercv#140] [Ideas]
A new Patent Entry would be helpful. It would be similar to the Publication Entry but would have the following fields:
- Status (Pending or Approved)
- Year submitted/accepted
- Application/Patent number

**Comment by sinaatalay** [comment]
RenderCV's entry types are not data-oriented but rendering-oriented. We would only want to create a new entry type if we require a new rendering type.

You can showcase your patents with one of the seven [available entry types](https://docs.rendercv.com/user_guide/structure_of_the_yaml_input_file/#cvsections-field). Maybe `NormalEntry` would be the best?

I am open to discussing any proposals for new rendering types if you have suggestions.

**Comment by elaske** [comment]
It's possible that this could be achieved through hijacking the `PublicationEntry` type as it has more in common with that. A link to the Application / Patent with similar functionality to the DOI field might work (e.g., render as a link to Google Patents or similar). 

If this can be achieved "simply" by overriding the rendering for `PublicationEntry` with an [arbitrary field](https://docs.rendercv.com/user_guide/structure_of_the_yaml_input_file/#using-arbitrary-keys), then it could be supported.

**Reply by elaske** [reply]
@sinaatalay I do think that patents likely should be treated slightly differently. I haven't tried the arbitrary field and override to get it right, yet, but the `PublicationEntry` type is not "quite" right. 

* Even if you just want to mark some of the important things, then `authors` would be required, which at least has a semantically equivalent `inventors` for patents, but I really don't necessarily care to take up space with the inventor list necessarily and `authors` can't be `None` or `[]`.
* `journal` could be used to denote the patent number, but it sort of sits alone
* `url` arguably could be used to link to the patent, but it renders as a raw URL. Could add this as a markdown rendered link within the `title`, but I think I'd rather have that for the patent number.

`date` is equivalent, but the context would be up to the user. `title` is equivalent.

Note also that the patent number may be multiple when a patent is awarded in multiple jurisdictions for the same invention (e.g., US, EU, CN) with multiple links all tied to the same invention.

**Reply by sinaatalay** [reply]
Thank you for sharing your thoughts.

First, let me clarify what I meant in my answer above (which I will also include in the developer guide's FAQ):

*When should we consider adding a new entry type to RenderCV?*

We should add a new entry type if and only if the proposed design of the entry type cannot be achieved using any of the existing entry types. This is because RenderCV's entry types are not designed to function as data models. Their purpose is not to store specific data but rather to determine how a given set of strings will appear in the PDF.

For example, JSON Resume follows a data-oriented approach.  In JSON Resume, each entry type acts as a data model specifically designed to store structured information. RenderCV takes a design-oriented approach for two reasons:
- There would be too many different data models that would ultimately look more or less the same in the PDF.
- It would be impossible to provide all the different data models people might need.

Therefore, we decided to create entry types solely for their design output.

Let's consider the PatentEntry example: should we create it?

If PatentEntry's design cannot be met by any of the existing seven entry types, then yes, we should consider adding it. However, if the current options are sufficient, we shouldn't add it. For example, you mentioned wanting to display the patent number as a clickable link—this is already possible using Markdown links. Also, you don't need to use PublicationEntry for patents. I believe NormalEntry would work perfectly in this case.

Can you share the design in your mind as a screenshot so that we can decide if it can be reproduced with one of the seven entry types?

**Reply by elaske** [reply]
It wasn't my original post, but I am actively working on porting my own resume to resumecv so my thoughts are evolving a bit as I try to figure out how to render things best. Here's what I currently have just using `PublicationEntry`, which as you note, might not be the right approach:

![image](https://github.com/user-attachments/assets/f5a4d3b7-2064-4010-a018-33ac9740fb8b)

One of my previous iterations looked like this: 
![image](https://github.com/user-attachments/assets/d38e2bef-57fa-4678-a71b-5ac3e76aa22f)

I will see if I can come up with something within the confines of the current structure. That being said, across the board as I was looking through the issues, a lot of questions point back to the data model context compared to others.  I should open up to a separate discussion, namely as you note

> JSON Resume follows a data-oriented approach.  In JSON Resume, each entry type acts as a data model specifically designed to store structured information. RenderCV takes a design-oriented approach...

I made some points [here](https://github.com/rendercv/rendercv/issues/164#issuecomment-2629075254) that get at where my, potentially misplaced, hopes for a tool like this are that are probably a separate discussion thread.

**Reply by sinaatalay** [reply]
Instead of using the `url` field, I would use a Markdown link in the title to make a link with an arbitrary placeholder.
